### PR TITLE
fix: redirect invalid state errors to site url

### PR DESCRIPTION
**Pull request by kangmingtay**
## What kind of change does this PR introduce?
* Currently, invalid state errors on the oauth callback endpoint are returned as json responses - this makes error handling impossible for the client since the callback is invoked by the oauth provider and leaves users with an unfriendly error message.
* This PR aims to address this by redirecting errors to the Site URL. We can't redirect errors to the redirect url specified in the `state` because the `state` might not be verified yet, which may lead to malicious redirects.
* This changes the response for these type of errors to redirect to the site URL instead of failing at the `/callback` endpoint.

### CODE REVIEW

**Comment by kangmingtay** [comment]
staticcheck errors are fixed in [this commit](https://github.com/supabase/auth/pull/1721/commits/66823d900b13b471b7bffc41c6588ed9b192a834)

File: internal/api/external_oauth.go [hunk 1]
```diff
@@ -43,7 +33,21 @@ func (a *API) loadFlowState(w http.ResponseWriter, r *http.Request) (context.Con
...
 	}
-	return a.loadExternalState(ctx, state)
+
+	var err error
+	ctx, err = a.loadExternalState(ctx, r)
+	if err != nil {
+		u, uerr := url.ParseRequestURI(a.config.SiteURL)
+		if uerr != nil {
+			return ctx, internalServerError("site url is improperly formatted").WithInternalError(uerr)
+		}
```
Comment (hf) [inline internal/api/external_oauth.go:43]: Ideally this is done at config time, and here you just assume it's valid.

File: internal/api/external_oauth.go [hunk 2]
```diff
@@ -43,7 +33,21 @@ func (a *API) loadFlowState(w http.ResponseWriter, r *http.Request) (context.Con
...
+
+		q := getErrorQueryString(err, utilities.GetRequestID(ctx), observability.GetLogEntry(r).Entry, u.Query())
+		u.RawQuery = q.Encode()
+
+		http.Redirect(w, r, u.String(), http.StatusFound)
```
Comment (hf) [inline internal/api/external_oauth.go:48]: I think the redirect here should be this: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303

File: internal/api/external_oauth.go (hunk 1 above)
Comment (kangmingtay) [reply internal/api/external_oauth.go:43]: will resolve this in a FLUP PR since it's being used in quite a few places

**Review by hf — APPROVED** [review]

**Pull request merged**